src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -0,0 +1,221 @@
+            var viewLines = TextView.TextViewLines;
+
+            foreach (var changedSpan in changedSpanCollection)
+            {
+                if (!viewLines.IntersectsBufferSpan(changedSpan))
+                {
+                    continue;
+                }
+
+                var tagSpans = TagAggregator.GetTags(changedSpan);
+                foreach (var tagMappingSpan in tagSpans)
+                {
+                    if (!ShouldDrawTag(changedSpan, tagMappingSpan))
+                    {
+                        continue;
+                    }
+
+                    var mappedPoint = GetMappedPoint(changedSpan, tagMappingSpan);
+
+                    // mappedPoint is known to not be null here because it is checked in the ShouldDrawTag method call.
+                    var lineNum = mappedPoint!.Value.GetContainingLine().LineNumber;
+
+                    // If the line does not have an associated tagMappingSpan and changedSpan, then add the first one.
+                    if (!map.TryGetValue(lineNum, out var value))
+                    {
+                        map.Add(lineNum, tagMappingSpan);
+                    }
+
+                    // If the map has a value and the value is not a syntax error, then rewrite the value in the map since
+                    // each line with errors will be ordered by importance
+                    if (value is not null && value.Tag.ErrorType is not PredefinedErrorTypeNames.SyntaxError)
+                    {
+                        map[lineNum] = tagMappingSpan;
+                    }
+                }
+            }
+
+            return map;
+        }
+
+        /// <summary>
+        /// Iterates through the mapping of line number to span and draws the diagnostic in the appropriate position on the screen,
+        /// as well as adding the tag to the adornment layer.
+        /// </summary>
+        protected override void AddAdornmentsToAdornmentLayer_CallOnlyOnUIThread(NormalizedSnapshotSpanCollection changedSpanCollection)
+        {
+            // this method should only run on UI thread as we do WPF here.
+            Contract.ThrowIfFalse(TextView.VisualElement.Dispatcher.CheckAccess());
+            if (changedSpanCollection.IsEmpty())
+            {
+                return;
+            }
+
+            var viewLines = TextView.TextViewLines;
+            var map = GetSpansOnEachLine(changedSpanCollection);
+            foreach (var (lineNum, tagMappingSpan) in map)
+            {
+                // Mapping the IMappingTagSpan back up to the TextView's visual snapshot to ensure there will
+                // be no adornments drawn on disjoint spans.
+                if (!TryMapToSingleSnapshotSpan(tagMappingSpan.Span, TextView.TextSnapshot, out var span))
+                {
+                    continue;
+                }
+
+                var geometry = viewLines.GetMarkerGeometry(span);
+                if (geometry != null)
+                {
```

Comment:
nit: your flow control is confusing at times.  it would be nice to be consistent between when you decide to continue, versus when you decide to put the code you want to run nested in the check.  For example, in this case, i think it would be more consistent to do:

```suggestion
                var geometry = viewLines.GetMarkerGeometry(span);
                if (geometry == null)
                    continue;
```

to match the check above.  this is also a bit easier to read as one doesn't need to scan down to see that ther'es no 'else' clause associated with this that does work.

Reply:
good point, I need to be more cognizant of that. changed